# 1. Import the Data Frame

In [1]:
# Import Libraries

import pandas as pd
import numpy as np
import os

In [2]:
# Import Original Data (Path)

path = r'/Users/ladyarmanoid/Documents/Data Analytics/CareerFoundry/Course/2- Data Immersion/Achievement 4/Instacart Basket Analysis/'

In [3]:
# Import ords_prods_merge

ords_prods_merge = pd.read_pickle(os.path.join(path,'02 Data','Prepared Data','ords_prods_merge.pkl'))

In [4]:
# Check the Shape

ords_prods_merge.shape

(32404859, 20)

In [5]:
# Check the Head

ords_prods_merge.head(10)

order_id  user_id  number_of_orders  orders_day_of_week  order_hour_of_day  \
0   2539329        1                 1                   2                  8   
1   2398795        1                 2                   3                  7   
2    473747        1                 3                   3                 12   
3   2254736        1                 4                   4                  7   
4    431534        1                 5                   4                 15   
5   3367565        1                 6                   2                  7   
6    550135        1                 7                   1                  9   
7   3108588        1                 8                   1                 14   
8   2295261        1                 9                   1                 16   
9   2550362        1                10                   4                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN       196.0                1.0        0.0   
1                    15.0       196.0                1.0        1.0   
2                    21.0       196.0                1.0        1.0   
3                    29.0       196.0                1.0        1.0   
4                    28.0       196.0                1.0        1.0   
5                    19.0       196.0                1.0        1.0   
6                    20.0       196.0                1.0        1.0   
7                    14.0       196.0                2.0        1.0   
8                     0.0       196.0                4.0        1.0   
9                    30.0       196.0                1.0        1.0   

  product_name  aisle_id  department_id  prices _merge    price_range_loc  \
0         Soda        77              7     9.0   both  Mid-range product   
1         Soda        77              7     9.0   both  Mid-range product   
2         Soda        77              7     9.0   both  Mid-range product   
3         Soda        77              7     9.0   both  Mid-range product   
4         Soda        77              7     9.0   both  Mid-range product   
5         Soda        77              7     9.0   both  Mid-range product   
6         Soda        77              7     9.0   both  Mid-range product   
7         Soda        77              7     9.0   both  Mid-range product   
8         Soda        77              7     9.0   both  Mid-range product   
9         Soda        77              7     9.0   both  Mid-range product   

      busiest_day     busiest_days busiest_period_of_day  max_order  \
0  Regularly busy   Regularly busy        Average orders         10   
1  Regularly busy  Least busy days        Average orders         10   
2  Regularly busy  Least busy days           Most orders         10   
3      Least busy  Least busy days        Average orders         10   
4      Least busy  Least busy days           Most orders         10   
5  Regularly busy   Regularly busy        Average orders         10   
6  Regularly busy     Busiest days           Most orders         10   
7  Regularly busy     Busiest days           Most orders         10   
8  Regularly busy     Busiest days           Most orders         10   
9      Least busy  Least busy days        Average orders         10   

   loyalty_flag  
0  New customer  
1  New customer  
2  New customer  
3  New customer  
4  New customer  
5  New customer  
6  New customer  
7  New customer  
8  New customer  
9  New customer

# 2. Group Data by 'department_id' -> Mean aggregation of 'number_of_orders'

1. Split the data into groups based on “department_id.”
2. Apply the agg() function to each group to obtain the mean values for the “order_number” column.

In [6]:
ords_prods_merge.groupby('department_id').agg({'number_of_orders': ['mean']})

number_of_orders
                          mean
department_id                 
1                    15.457838
2                    17.277920
3                    17.170395
4                    17.811403
5                    15.215751
6                    16.439806
7                    17.225802
8                    15.340650
9                    15.895474
10                   20.197148
11                   16.170638
12                   15.887671
13                   16.583536
14                   16.773669
15                   16.165037
16                   17.665606
17                   15.694469
18                   19.310397
19                   17.177343
20                   16.473447
21                   22.902379

# 3. Analyze the results

In [8]:
# Import departments data frame

df_dept = pd.read_csv(os.path.join(path,'02 Data','Prepared Data','departments_wrangled.csv'))

In [9]:
# Check the output

df_dept

Unnamed: 0       department
0            1           frozen
1            2            other
2            3           bakery
3            4          produce
4            5          alcohol
5            6    international
6            7        beverages
7            8             pets
8            9  dry goods pasta
9           10             bulk
10          11    personal care
11          12     meat seafood
12          13           pantry
13          14        breakfast
14          15     canned goods
15          16       dairy eggs
16          17        household
17          18           babies
18          19           snacks
19          20             deli
20          21          missing

Compared to the subset used in the GL_Grouping_Data_Aggregating_Variables_Practice_4.8, we find 21 departments instead of 8 (limited by the 1.000.000 rows).

The majority of the mean values are close to the one the subset results gave:

- The minimum mean order is 11.3 in the subset while it is 15.2 in the dataframe
- The maximum mean is 19.5 in the subset and 20.1 in the dataframe (not including department 21, the "missing" category)
- The department ID with the largest number of average orders is that of the "missing" category, indicating a large number of unspecified values for the column "department_id."
- The next largest number of average orders is the "bulk" department, which was not avalaible in the subset.

# 4. Create a Loyalty Flag for existing customers

In [10]:
# Create new column 'max_order'

ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['number_of_orders'].transform(np.max)

/var/folders/bx/5gbvgdk148552knvq9wj8vgm0000gn/T/ipykernel_8005/1007047095.py:3: FutureWarning: The provided callable <function amax at 0x10e030a40> is currently using SeriesGroupBy.max. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "max" instead.
  ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['number_of_orders'].transform(np.max)


In [11]:
# Check the new column

ords_prods_merge.head(20)

order_id  user_id  number_of_orders  orders_day_of_week  \
0    2539329        1                 1                   2   
1    2398795        1                 2                   3   
2     473747        1                 3                   3   
3    2254736        1                 4                   4   
4     431534        1                 5                   4   
5    3367565        1                 6                   2   
6     550135        1                 7                   1   
7    3108588        1                 8                   1   
8    2295261        1                 9                   1   
9    2550362        1                10                   4   
10   2968173       15                15                   1   
11   1870022       15                17                   2   
12   1911383       15                18                   2   
13   2715276       15                21                   1   
14    487368       15                22                   1   
15   2293453       19                 2                   5   
16   1973799       19                 5                   6   
17    532817       19                 7                   4   
18   1573906       21                10                   3   
19   1593000       31                10                   3   

    order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                   8                     NaN       196.0                1.0   
1                   7                    15.0       196.0                1.0   
2                  12                    21.0       196.0                1.0   
3                   7                    29.0       196.0                1.0   
4                  15                    28.0       196.0                1.0   
5                   7                    19.0       196.0                1.0   
6                   9                    20.0       196.0                1.0   
7                  14                    14.0       196.0                2.0   
8                  16                     0.0       196.0                4.0   
9                   8                    30.0       196.0                1.0   
10                  9                     7.0       196.0                2.0   
11                 16                     8.0       196.0                6.0   
12                 11                     7.0       196.0                1.0   
13                  9                     7.0       196.0                1.0   
14                 10                    14.0       196.0                1.0   
15                 14                     6.0       196.0                3.0   
16                 12                     8.0       196.0               15.0   
17                 17                     6.0       196.0                1.0   
18                 10                     6.0       196.0                2.0   
19                  8                     7.0       196.0               17.0   

    reordered product_name  aisle_id  department_id  prices _merge  \
0         0.0         Soda        77              7     9.0   both   
1         1.0         Soda        77              7     9.0   both   
2         1.0         Soda        77              7     9.0   both   
3         1.0         Soda        77              7     9.0   both   
4         1.0         Soda        77              7     9.0   both   
5         1.0         Soda        77              7     9.0   both   
6         1.0         Soda        77              7     9.0   both   
7         1.0         Soda        77              7     9.0   both   
8         1.0         Soda        77              7     9.0   both   
9         1.0         Soda        77              7     9.0   both   
10        0.0         Soda        77              7     9.0   both   
11        1.0         Soda        77              7     9.0   both   
12        1.0         Soda        77              7     9.0   both   
13        1.0      

In [12]:
# Loyalty Flag 'Loyal Customer'

ords_prods_merge.loc[ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'

In [13]:
# Loyalty Flag 'Regular Customer'

ords_prods_merge.loc[(ords_prods_merge['max_order'] <= 40) & (ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'

In [14]:
# Loyalty Flag 'New Customer'

ords_prods_merge.loc[ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [15]:
# Check Loyalty Flags

ords_prods_merge['loyalty_flag'].value_counts(dropna = False)

loyalty_flag
Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: count, dtype: int64

In [16]:
# Check the output -> Column 'Loyalty Flag'

ords_prods_merge[['user_id', 'loyalty_flag', 'number_of_orders']].head(20)

user_id      loyalty_flag  number_of_orders
0         1      New customer                 1
1         1      New customer                 2
2         1      New customer                 3
3         1      New customer                 4
4         1      New customer                 5
5         1      New customer                 6
6         1      New customer                 7
7         1      New customer                 8
8         1      New customer                 9
9         1      New customer                10
10       15  Regular customer                15
11       15  Regular customer                17
12       15  Regular customer                18
13       15  Regular customer                21
14       15  Regular customer                22
15       19      New customer                 2
16       19      New customer                 5
17       19      New customer                 7
18       21  Regular customer                10
19       31  Regular customer                10

# 5. Difference in spending habits for the three types of customers

In [17]:
# Group loyalty flag customers by the min, max, and mean prices

ords_prods_merge.groupby('loyalty_flag').agg({'prices': ['mean', 'min', 'max']})

prices              
                       mean  min      max
loyalty_flag                             
Loyal customer    10.386336  1.0  99999.0
New customer      13.294670  1.0  99999.0
Regular customer  12.495717  1.0  99999.0

Loyal customers are spending the least, while new customers are spending the most.
The regular customers are in between.

There is a clear outlier for the maximum price value.
For this reason, we have to have them marked as missing value.

In [18]:
# Mark outliers as missing data

ords_prods_merge.loc[ords_prods_merge['prices'] >100, 'prices'] = np.nan

In [19]:
# Check the output

ords_prods_merge['prices'].max()

25.0

In [20]:
# Group loyalty flag customers by the min, max, and mean prices

ords_prods_merge.groupby('loyalty_flag').agg({'prices': ['mean', 'min', 'max']})

prices           
                      mean  min   max
loyalty_flag                         
Loyal customer    7.773575  1.0  25.0
New customer      7.801206  1.0  25.0
Regular customer  7.798262  1.0  25.0

# 6. Create a Spending Flag for users based on the avg. price basket

In [21]:
# Create average_spend column

ords_prods_merge['average_spend'] = ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)

/var/folders/bx/5gbvgdk148552knvq9wj8vgm0000gn/T/ipykernel_8005/1921476679.py:3: FutureWarning: The provided callable <function mean at 0x10e031760> is currently using SeriesGroupBy.mean. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "mean" instead.
  ords_prods_merge['average_spend'] = ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)


In [22]:
# Checking the new column

ords_prods_merge.head(20)

order_id  user_id  number_of_orders  orders_day_of_week  \
0    2539329        1                 1                   2   
1    2398795        1                 2                   3   
2     473747        1                 3                   3   
3    2254736        1                 4                   4   
4     431534        1                 5                   4   
5    3367565        1                 6                   2   
6     550135        1                 7                   1   
7    3108588        1                 8                   1   
8    2295261        1                 9                   1   
9    2550362        1                10                   4   
10   2968173       15                15                   1   
11   1870022       15                17                   2   
12   1911383       15                18                   2   
13   2715276       15                21                   1   
14    487368       15                22                   1   
15   2293453       19                 2                   5   
16   1973799       19                 5                   6   
17    532817       19                 7                   4   
18   1573906       21                10                   3   
19   1593000       31                10                   3   

    order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                   8                     NaN       196.0                1.0   
1                   7                    15.0       196.0                1.0   
2                  12                    21.0       196.0                1.0   
3                   7                    29.0       196.0                1.0   
4                  15                    28.0       196.0                1.0   
5                   7                    19.0       196.0                1.0   
6                   9                    20.0       196.0                1.0   
7                  14                    14.0       196.0                2.0   
8                  16                     0.0       196.0                4.0   
9                   8                    30.0       196.0                1.0   
10                  9                     7.0       196.0                2.0   
11                 16                     8.0       196.0                6.0   
12                 11                     7.0       196.0                1.0   
13                  9                     7.0       196.0                1.0   
14                 10                    14.0       196.0                1.0   
15                 14                     6.0       196.0                3.0   
16                 12                     8.0       196.0               15.0   
17                 17                     6.0       196.0                1.0   
18                 10                     6.0       196.0                2.0   
19                  8                     7.0       196.0               17.0   

    reordered product_name  ...  department_id  prices  _merge  \
0         0.0         Soda  ...              7     9.0    both   
1         1.0         Soda  ...              7     9.0    both   
2         1.0         Soda  ...              7     9.0    both   
3         1.0         Soda  ...              7     9.0    both   
4         1.0         Soda  ...              7     9.0    both   
5         1.0         Soda  ...              7     9.0    both   
6         1.0         Soda  ...              7     9.0    both   
7         1.0         Soda  ...              7     9.0    both   
8         1.0         Soda  ...              7     9.0    both   
9         1.0         Soda  ...              7     9.0    both   
10        0.0         Soda  ...              7     9.0    both   
11        1.0         Soda  ...              7     9.0    both   
12        1.0         Soda  ...              7     9.0    both   
13        1.0         Soda  ...              7     9.0    both   
14       

In [23]:
# Flag for 'Low Spender'

ords_prods_merge.loc[ords_prods_merge['average_spend']<10,'spender_flag']='low_spender'

In [24]:
# Flag for 'High Spender'

ords_prods_merge.loc[ords_prods_merge['average_spend']>= 10, 'spender_flag'] = 'high_spender'

In [25]:
# Check Spender Flags

ords_prods_merge['spender_flag'].value_counts(dropna = False)

spender_flag
low_spender     32285131
high_spender      119728
Name: count, dtype: int64

In [26]:
# Check the output

ords_prods_merge[['user_id', 'average_spend', 'spender_flag', 'prices']].head(60)

user_id  average_spend spender_flag  prices
0         1       6.367797  low_spender     9.0
1         1       6.367797  low_spender     9.0
2         1       6.367797  low_spender     9.0
3         1       6.367797  low_spender     9.0
4         1       6.367797  low_spender     9.0
5         1       6.367797  low_spender     9.0
6         1       6.367797  low_spender     9.0
7         1       6.367797  low_spender     9.0
8         1       6.367797  low_spender     9.0
9         1       6.367797  low_spender     9.0
10       15       3.980556  low_spender     9.0
11       15       3.980556  low_spender     9.0
12       15       3.980556  low_spender     9.0
13       15       3.980556  low_spender     9.0
14       15       3.980556  low_spender     9.0
15       19       8.372059  low_spender     9.0
16       19       8.372059  low_spender     9.0
17       19       8.372059  low_spender     9.0
18       21       7.012683  low_spender     9.0
19       31       8.137793  low_spender     9.0
20       31       8.137793  low_spender     9.0
21       43       7.444898  low_spender     9.0
22       43       7.444898  low_spender     9.0
23       52       6.356287  low_spender     9.0
24       52       6.356287  low_spender     9.0
25       52       6.356287  low_spender     9.0
26       52       6.356287  low_spender     9.0
27       52       6.356287  low_spender     9.0
28       52       6.356287  low_spender     9.0
29       52       6.356287  low_spender     9.0
30       52       6.356287  low_spender     9.0
31       52       6.356287  low_spender     9.0
32       52       6.356287  low_spender     9.0
33       52       6.356287  low_spender     9.0
34       52       6.356287  low_spender     9.0
35       52       6.356287  low_spender     9.0
36       52       6.356287  low_spender     9.0
37       67       7.656790  low_spender     9.0
38       67       7.656790  low_spender     9.0
39       67       7.656790  low_spender     9.0
40       67       7.656790  low_spender     9.0
41       67       7.656790  low_spender     9.0
42       67       7.656790  low_spender     9.0
43       67       7.656790  low_spender     9.0
44       67       7.656790  low_spender     9.0
45       67       7.656790  low_spender     9.0
46       67       7.656790  low_spender     9.0
47       67       7.656790  low_spender     9.0
48       67       7.656790  low_spender     9.0
49       67       7.656790  low_spender     9.0
50       67       7.656790  low_spender     9.0
51       67       7.656790  low_spender     9.0
52       67       7.656790  low_spender     9.0
53       67       7.656790  low_spender     9.0
54       67       7.656790  low_spender     9.0
55       67       7.656790  low_spender     9.0
56       81       8.165385  low_spender     9.0
57       81       8.165385  low_spender     9.0
58       82       6.441221  low_spender     9.0
59       82       6.441221  low_spender     9.0

# 7. Create a Flag for Frequent Users

In [27]:
# Create a column for Frequent Users
# Group data by 'user_id'
# Generate median product of 'days_since_prior_order'
# Create 'customer_frequency' to aggregate the results


ords_prods_merge['customer_frequency'] = ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform(np.median)

/var/folders/bx/5gbvgdk148552knvq9wj8vgm0000gn/T/ipykernel_8005/1702137845.py:7: FutureWarning: The provided callable <function median at 0x10e16b9c0> is currently using SeriesGroupBy.median. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "median" instead.
  ords_prods_merge['customer_frequency'] = ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform(np.median)


In [28]:
# Check the new column

ords_prods_merge.head(20)

order_id  user_id  number_of_orders  orders_day_of_week  \
0    2539329        1                 1                   2   
1    2398795        1                 2                   3   
2     473747        1                 3                   3   
3    2254736        1                 4                   4   
4     431534        1                 5                   4   
5    3367565        1                 6                   2   
6     550135        1                 7                   1   
7    3108588        1                 8                   1   
8    2295261        1                 9                   1   
9    2550362        1                10                   4   
10   2968173       15                15                   1   
11   1870022       15                17                   2   
12   1911383       15                18                   2   
13   2715276       15                21                   1   
14    487368       15                22                   1   
15   2293453       19                 2                   5   
16   1973799       19                 5                   6   
17    532817       19                 7                   4   
18   1573906       21                10                   3   
19   1593000       31                10                   3   

    order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                   8                     NaN       196.0                1.0   
1                   7                    15.0       196.0                1.0   
2                  12                    21.0       196.0                1.0   
3                   7                    29.0       196.0                1.0   
4                  15                    28.0       196.0                1.0   
5                   7                    19.0       196.0                1.0   
6                   9                    20.0       196.0                1.0   
7                  14                    14.0       196.0                2.0   
8                  16                     0.0       196.0                4.0   
9                   8                    30.0       196.0                1.0   
10                  9                     7.0       196.0                2.0   
11                 16                     8.0       196.0                6.0   
12                 11                     7.0       196.0                1.0   
13                  9                     7.0       196.0                1.0   
14                 10                    14.0       196.0                1.0   
15                 14                     6.0       196.0                3.0   
16                 12                     8.0       196.0               15.0   
17                 17                     6.0       196.0                1.0   
18                 10                     6.0       196.0                2.0   
19                  8                     7.0       196.0               17.0   

    reordered product_name  ...  _merge    price_range_loc     busiest_day  \
0         0.0         Soda  ...    both  Mid-range product  Regularly busy   
1         1.0         Soda  ...    both  Mid-range product  Regularly busy   
2         1.0         Soda  ...    both  Mid-range product  Regularly busy   
3         1.0         Soda  ...    both  Mid-range product      Least busy   
4         1.0         Soda  ...    both  Mid-range product      Least busy   
5         1.0         Soda  ...    both  Mid-range product  Regularly busy   
6         1.0         Soda  ...    both  Mid-range product  Regularly busy   
7         1.0         Soda  ...    both  Mid-range product  Regularly busy   
8         1.0         Soda  ...    both  Mid-range product  Regularly busy   
9         1.0         Soda  ...    both  Mid-range product      Least busy   
10        0.0         Soda  ...    both  Mid-range product  Regularly busy   
11        1.0         Soda  ...    both  Mid-range product  Reg

In [29]:
# Flag for 'Non-Frequent Customer'

ords_prods_merge.loc[ords_prods_merge['customer_frequency'] > 20, 'frequency_flag'] = 'non_frequent_customer'

In [30]:
# Flag for 'Regular Customer'

ords_prods_merge.loc[(ords_prods_merge['customer_frequency']>10) & (ords_prods_merge['customer_frequency']<=20), 'frequency_flag'] = 'regular_customer'

In [31]:
# Flag for 'Frequent Customer'

ords_prods_merge.loc[ords_prods_merge['customer_frequency'] <= 10, 'frequency_flag'] = 'frequent_customer'

In [32]:
# Check Frequency Flags

ords_prods_merge['frequency_flag'].value_counts(dropna = False)

frequency_flag
frequent_customer        21559853
regular_customer          7208564
non_frequent_customer     3636437
nan                             5
Name: count, dtype: int64

In [33]:
#Check new output

ords_prods_merge[['user_id', 'customer_frequency', 'frequency_flag']].head(60)

user_id  customer_frequency         frequency_flag
0         1                20.5  non_frequent_customer
1         1                20.5  non_frequent_customer
2         1                20.5  non_frequent_customer
3         1                20.5  non_frequent_customer
4         1                20.5  non_frequent_customer
5         1                20.5  non_frequent_customer
6         1                20.5  non_frequent_customer
7         1                20.5  non_frequent_customer
8         1                20.5  non_frequent_customer
9         1                20.5  non_frequent_customer
10       15                10.0      frequent_customer
11       15                10.0      frequent_customer
12       15                10.0      frequent_customer
13       15                10.0      frequent_customer
14       15                10.0      frequent_customer
15       19                 6.0      frequent_customer
16       19                 6.0      frequent_customer
17       19                 6.0      frequent_customer
18       21                 7.5      frequent_customer
19       31                 6.0      frequent_customer
20       31                 6.0      frequent_customer
21       43                10.0      frequent_customer
22       43                10.0      frequent_customer
23       52                 8.0      frequent_customer
24       52                 8.0      frequent_customer
25       52                 8.0      frequent_customer
26       52                 8.0      frequent_customer
27       52                 8.0      frequent_customer
28       52                 8.0      frequent_customer
29       52                 8.0      frequent_customer
30       52                 8.0      frequent_customer
31       52                 8.0      frequent_customer
32       52                 8.0      frequent_customer
33       52                 8.0      frequent_customer
34       52                 8.0      frequent_customer
35       52                 8.0      frequent_customer
36       52                 8.0      frequent_customer
37       67                 7.0      frequent_customer
38       67                 7.0      frequent_customer
39       67                 7.0      frequent_customer
40       67                 7.0      frequent_customer
41       67                 7.0      frequent_customer
42       67                 7.0      frequent_customer
43       67                 7.0      frequent_customer
44       67                 7.0      frequent_customer
45       67                 7.0      frequent_customer
46       67                 7.0      frequent_customer
47       67                 7.0      frequent_customer
48       67                 7.0      frequent_customer
49       67                 7.0      frequent_customer
50       67                 7.0      frequent_customer
51       67                 7.0      frequent_customer
52       67                 7.0      frequent_customer
53       67                 7.0      frequent_customer
54       67                 7.0      frequent_customer
55       67                 7.0      frequent_customer
56       81                30.0  non_frequent_customer
57       81                30.0  non_frequent_customer
58       82                11.5       regular_customer
59       82                11.5       regular_customer

In [34]:
# Exporting ords_prods_merge using PKL

ords_prods_merge.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'ords_prods_aggregated.pkl'))